jsonnet/prometheus-adapter.libsonnet

```diff
@@ -59,6 +59,8 @@ function(params)
               caFile: '/etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt',
               serverName: 'server-name-replaced-at-runtime',
               insecureSkipVerify: false,
+              // TODO: prometheus-adapter currently is a stock upstream aggregated api server.
+              // It does not support static authorization.
```

Comment:
Do we have an upstream issue against kube-apiserver to support static authorization for the aggregation layer? As far as I can tell from the [official doc](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/#authentication-flow) detailing the authentication flow for aggregated APIs, the authorization between the apiserver and the aggregated API can currently only be made by submitting a `SubjectAccessReview` call.

Reply:
I believe sig-auth has already discussed static authorization in the past, although I am not sure it was based upon a specific issue.

I am not aware of this being in the upstream works today, but in OpenShift we've got this implemented in the operators that are based on library-go: https://github.com/openshift/library-go/pull/998

Reply:
Awesome, that's exactly what I was looking for, thanks @stlaz.

If I understand things correctly, https://github.com/openshift/kubernetes/pull/588 added static authorization to the apiserver, so theoretically we should be able to use it in prometheus-adapter. But, thinking about it again, authorizing custom resources such as PodMetrics and NodeMetrics might be out of scope since I don't see them in the goals of the OEP.

On another note, this discussion made me realize that prometheus-adapter isn't protected by a kube-rbac-proxy even though we are scraping its metrics. I will open a separate PR to fix that.

Reply:
I encountered an issue while adding authz to prometheus-adapter metrics endpoints. Since metrics serving isn't decoupled from the generic apiserver code, scrapers need to be authorized by the apiserver in order to access the server. So if we add a kube-rbac-proxy in front of the /metrics endpoint of prometheus-adapter, the proxy would need to forward the bearer token in order for the scrape to succeed. But, kube-rbac-proxy wasn't designed to allow that and is only supposed to forward requests to applications that aren't kubernetes aware.

I opened https://github.com/kubernetes-sigs/prometheus-adapter/issues/425 upstream.

Reply:
The decoupling would be one way to do it, another would be to have the static authorizer be a part of the prometheus-adapter.

This PR attempts to make it possible to use client certificate authentication for the scraping. For that, the side receiving the scraping request only needs to have the correct CA bundle and would not have to send any requests to the kube-apiserver.

Please know that we're not aiming for a 100% coverage of static authn/z from the beginning. For now it would be enough if we could get this PR merged, and then we'll have to update the ServiceMonitors for most of our components so that the client cert authentication is taken into effect.

I'll check whether the e2e test failures are related to this PR.

Reply:
Yeah, it is definitely not a blocker for me, I just wanted to make sure that we at least track this effort.